File: java/binary-search/src/main/java/BinarySearch.java
import java.util.List;

class BinarySearch {

    private List<Integer> items;

    BinarySearch(List<Integer> items) {
        this.items = items;
    }

    int indexOf(int item) throws ValueNotFoundException {
        final int itemIndex = this.find(item, 0, this.items.size() - 1);

        if (itemIndex == -1) {
            throw new ValueNotFoundException("Value not in array");
        }

        return itemIndex;
    }

    private int find(int item, int bot, int top) {
        if (bot > top) {
            return -1;
        }

        final int middle = (bot + top) / 2;
        final int itemInMiddle = this.items.get(middle);

        if (itemInMiddle < item) {
            return this.find(item, middle + 1, top);
        }
        if (itemInMiddle > item) {
            return this.find(item, bot, middle - 1);

        }
        return middle;
    }

}

File: go/binary-search/binary_search.go
// Package binarysearch provides functions for performing binary search on sorted data.
package binarysearch

// SearchInts searches for an integer key in a sorted slice of integers.
// It returns the index of the key if found, or -1 if the key is not present.
// The provided slice must be sorted in ascending order for the search to work correctly.
func SearchInts(list []int, key int) int {
	for bot, top := 0, len(list)-1; bot <= top; {
		middle := (bot + top) / 2
		candidate := list[middle]

		if candidate == key {
			return middle
		}

		if candidate > key {
			top = middle - 1
		} else {
			bot = middle + 1
		}
	}
	return -1
}
